src/ZeroLog/EnumCache.cs

```diff
@@ -0,0 +1,200 @@
+            {
+                switch (Type.GetTypeCode(Enum.GetUnderlyingType(value.GetType())))
+                {
+                    case TypeCode.Byte:
+                        return (byte)(object)value;
+
+                    case TypeCode.SByte:
+                        return (ulong)(sbyte)(object)value;
+
+                    case TypeCode.Int16:
+                        return (ulong)(short)(object)value;
+
+                    case TypeCode.UInt16:
+                        return (ushort)(object)value;
+
+                    case TypeCode.Int32:
+                        return (ulong)(int)(object)value;
+
+                    case TypeCode.UInt32:
+                        return (uint)(object)value;
+
+                    case TypeCode.Int64:
+                        return (ulong)(long)(object)value;
+
+                    case TypeCode.UInt64:
+                        return (ulong)(object)value;
+
+                    default:
+                        throw new InvalidOperationException($"Invalid enum: {value.GetType()}");
+                }
+            }
+        }
+
+        [SuppressMessage("ReSharper", "ConvertClosureToMethodGroup")]
+        public static bool IsEnumSigned(IntPtr typeHandle)
+        {
+            return _isEnumSigned.GetOrAdd(typeHandle, h => IsEnumSignedImpl(h));
+
+            bool IsEnumSignedImpl(IntPtr h)
+            {
+                var type = TypeUtil.GetTypeFromHandle(h);
+                switch (Type.GetTypeCode(Enum.GetUnderlyingType(type)))
+                {
+                    case TypeCode.SByte:
+                    case TypeCode.Int16:
+                    case TypeCode.Int32:
+                    case TypeCode.Int64:
+                        return true;
+
+                    case TypeCode.Byte:
+                    case TypeCode.UInt16:
+                    case TypeCode.UInt32:
+                    case TypeCode.UInt64:
+                        return false;
+
+                    default:
+                        throw new InvalidOperationException($"Invalid enum: {type}");
+                }
+            }
+        }
+
+        private abstract class EnumStrings
+        {
+            public static EnumStrings Create(Type enumType)
+            {
+                var enumItems = Enum.GetValues(enumType)
+                                    .Cast<Enum>()
+                                    .Select(i => new EnumItem(i))
+                                    .ToList();
+
+                return ArrayEnumStrings.CanHandle(enumItems)
+                    ? (EnumStrings)new ArrayEnumStrings(enumItems)
+                    : new DictionaryEnumStrings(enumItems);
+            }
+
+            [CanBeNull]
+            public abstract string TryGetString(ulong value);
+        }
+
+        private sealed class ArrayEnumStrings : EnumStrings
+        {
+            private readonly string[] _strings;
+
+#if NET452
+            private static readonly string[] _emptyStrings = new string[0];
```

Comment:
This just creates noise in the code and leaves as much (or more) code to be migrated when we no longer support .NET 4.5

Reply:
Yes, that's unfortunate. I'll reimplement `Array.Empty<T>` as an internal utility.